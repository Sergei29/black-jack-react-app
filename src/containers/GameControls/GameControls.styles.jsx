import styled from "styled-components";

export const HitButtonContainer = styled.div`
	width: 100%;
	height: 200px;
	margin: 0 auto;
	display: flex;
	justify-content: center;
	align-items: center;
	background-image: url("https://upload.wikimedia.org/wikipedia/commons/thumb/9/95/A_face-up_deck_of_cards.jpg/1280px-A_face-up_deck_of_cards.jpg");
	background-size: cover;
	button {
		width: 150px;
		height: 150px;
		background-color: rgba(0, 0, 0, 0.5);
		border-radius: 50%;
		color: #fff;
		font-weight: bold;
		font-size: 28px;
		&:focus {
			outline: none;
		}
		&:active {
			color: red;
		}
	}
`;

export const CurrentStatusContainer = styled.div`
	p {
		text-align: center;
	}
`;

export const ControlButtonsContainer = styled.div`
	display: flex;
	justify-content: center;
	button {
		border-radius: 0;
		padding: 6px 12px;
		margin: 12px;
		font-size: 21px;
		font-weight: bold;
	}
`;
