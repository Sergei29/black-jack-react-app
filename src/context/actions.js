import { deck } from "../data/gameData";
import actionsTypes from "./actions.types";
import {
	getInitialHands,
	drawCardForPlayer,
	drawCardsForDealer,
	findTheWinner,
} from "./utils";

export const resetGame = () => {
	const newState = {
		playerDone: false,
		dealerDone: false,
		winner: null,
	};
	const { newDeck, newDealer, newPlayer } = getInitialHands(deck);

	newState.deck = newDeck;
	newState.dealer = newDealer;
	newState.player = newPlayer;

	return {
		type: actionsTypes.RESET_GAME,
		payload: newState,
	};
};

export const cardForPlayer = (deck, player) => {
	const { newDeck, newPlayer } = drawCardForPlayer(deck, player);
	return {
		type: actionsTypes.CARD_FOR_PLAYER,
		payload: { newDeck, newPlayer },
	};
};

export const isPlayerDone = (player) => ({
	type: actionsTypes.IS_PLAYER_DONE,
	payload: player.total > 21,
});

export const playerIsDone = () => ({
	type: actionsTypes.PLAYER_IS_DONE,
});

export const cardsForDealer = (deck, dealer, player) => {
	const { newDeck, newDealer } = drawCardsForDealer(deck, dealer, player);
	return {
		type: actionsTypes.CARDS_FOR_DEALER,
		payload: { newDeck, newDealer },
	};
};

export const findWinner = (dealer, player) => {
	const newWinner = findTheWinner(dealer, player);
	return {
		type: actionsTypes.FIND_WINNER,
		payload: newWinner,
	};
};
