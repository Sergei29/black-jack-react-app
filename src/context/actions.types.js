const actionsTypes = {
	RESET_GAME: "RESET_GAME",
	CARD_FOR_PLAYER: "CARD_FOR_PLAYER",
	IS_PLAYER_DONE: "IS_PLAYER_DONE",
	PLAYER_IS_DONE: "PLAYER_IS_DONE",
	DRAW_CARDS_FOR_DEALER: "DRAW_CARDS_FOR_DEALER",
	FIND_WINNER: "FIND_WINNER",
};

export default actionsTypes;
