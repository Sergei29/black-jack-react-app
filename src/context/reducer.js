import actionsTypes from "./actions.types";

export const INITIAL_STATE = {
	deck: [],
	player: { cards: [], total: 0 },
	dealer: { cards: [], total: 0 },
	playerDone: false,
	dealerDone: false,
	winner: null,
};

export const reducer = (state = INITIAL_STATE, action) => {
	switch (action.type) {
		case actionsTypes.RESET_GAME:
			return action.payload;

		case actionsTypes.CARD_FOR_PLAYER:
			return {
				...state,
				deck: action.payload.newDeck,
				player: action.payload.newPlayer,
			};

		case actionsTypes.IS_PLAYER_DONE:
			return {
				...state,
				playerDone: action.payload,
			};

		case actionsTypes.PLAYER_IS_DONE:
			return {
				...state,
				playerDone: true,
			};

		case actionsTypes.CARDS_FOR_DEALER:
			return {
				...state,
				deck: action.payload.newDeck,
				dealer: action.payload.newDealer,
				dealerDone: true,
			};

		case actionsTypes.FIND_WINNER:
			return {
				...state,
				winner: action.payload,
			};

		default:
			return state;
	}
};
