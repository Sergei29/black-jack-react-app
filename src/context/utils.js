import { roles } from "../data/gameData";

export const shuffleDeck = (deck) => [...deck].sort(() => Math.random() - 0.5);

export const calculateTotal = (currentTotal, card) => {
	let score = 0;
	if (!isNaN(+card.rank)) {
		score = +card.rank;
	} else if (card.rank === "A") {
		score = currentTotal + 11 > 21 ? 1 : 11;
	} else {
		score = 10;
	}
	return currentTotal + score;
};

export const getInitialHands = (deck) => {
	const newDeck = shuffleDeck([...deck]);
	const newPlayer = { cards: [], total: 0 };
	const newDealer = { cards: [], total: 0 };

	// first: 2 cards for player:
	for (let i = 0; i < 2; i++) {
		const card = newDeck.pop();
		newPlayer.cards.push(card);
		newPlayer.total = calculateTotal(newPlayer.total, card);
	}

	// then: 1 card for dealer:
	const cardForDealer = newDeck.pop();
	newDealer.cards.push(cardForDealer);
	newDealer.total = calculateTotal(newDealer.total, cardForDealer);

	return {
		newDeck,
		newPlayer,
		newDealer,
	};
};

export const drawCardForPlayer = (currentDeck, player) => {
	const newPlayer = {
		...player,
		cards: [...player.cards],
	};
	const newDeck = [...currentDeck];
	const newCard = newDeck.pop();
	newPlayer.cards.push(newCard);
	newPlayer.total = calculateTotal(player.total, newCard);

	return { newDeck, newPlayer };
};

export const drawCardsForDealer = (deck, dealer, player) => {
	let newDealer = {
		...dealer,
		cards: [...dealer.cards],
	};
	const newDeck = [...deck];

	if (player.total > 21) {
		return { newDeck, newDealer };
	}

	while (newDealer.total < player.total) {
		const card = newDeck.pop();
		newDealer.cards.push(card);
		newDealer.total = calculateTotal(newDealer.total, card);
	}

	return { newDeck, newDealer };
};

export const findTheWinner = (dealer, player) => {
	let newWinner = null;

	if (dealer.total > 21) return roles.player;
	if (player.total > 21) return roles.dealer;

	if (dealer.total >= player.total) {
		newWinner = roles.dealer;
	} else {
		newWinner = roles.player;
	}
	return newWinner;
};
