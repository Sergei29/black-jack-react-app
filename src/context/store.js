import React, { createContext, useReducer } from "react";
import { INITIAL_STATE, reducer } from "./reducer";

export const store = createContext(INITIAL_STATE);
const { Provider } = store;

export const StoreProvider = ({ children }) => {
	const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
	return <Provider value={{ state, dispatch }}> {children} </Provider>;
};
