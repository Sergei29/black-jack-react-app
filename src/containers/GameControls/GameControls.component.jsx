import React, { useContext, useEffect } from "react";
import { store } from "../../context/store";
import {
	resetGame,
	cardForPlayer,
	isPlayerDone,
	playerIsDone,
	cardsForDealer,
	findWinner,
} from "../../context/actions";
//components:
import StatusDisplay from "../../components/StatusDisplay/StatusDisplay.component";

//styles:
import {
	ControlButtonsContainer,
	HitButtonContainer,
} from "./GameControls.styles";

const GameControls = () => {
	const { state, dispatch } = useContext(store);
	const { deck, dealer, player, playerDone, dealerDone, winner } = state;

	useEffect(() => {
		dispatch(resetGame());
	}, [dispatch]);

	useEffect(() => {
		if (!playerDone) {
			dispatch(isPlayerDone(player));
		}
	}, [deck]);

	useEffect(() => {
		if (playerDone) {
			dispatch(cardsForDealer(deck, dealer, player));
		}
	}, [playerDone]);

	useEffect(() => {
		if (dealerDone) {
			dispatch(findWinner(dealer, player));
		}
	}, [dealerDone]);

	const hitClickHandler = () => {
		dispatch(cardForPlayer(deck, player));
	};

	const stickClickHandler = () => {
		dispatch(playerIsDone());
	};

	const resetClickHandler = () => {
		dispatch(resetGame());
	};

	return (
		<>
			<HitButtonContainer>
				<button onClick={hitClickHandler} disabled={playerDone}>
					Hit
				</button>
			</HitButtonContainer>
			<ControlButtonsContainer>
				<button onClick={stickClickHandler} disabled={playerDone}>
					Stick
				</button>

				<button onClick={resetClickHandler} disabled={!playerDone}>
					Reset Game
				</button>
			</ControlButtonsContainer>
			<StatusDisplay
				dealerTotal={dealer.total}
				playerTotal={player.total}
				winner={winner}
			/>
		</>
	);
};

export default GameControls;
