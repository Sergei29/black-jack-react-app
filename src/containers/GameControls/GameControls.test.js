import React from "react";
import { mount } from "enzyme";
import { StoreProvider } from "../../context/store";
import GameControls from "./GameControls.component";

describe("GameControls", () => {
	let wrapper;

	beforeEach(() => {
		wrapper = mount(
			<StoreProvider>
				<GameControls />
			</StoreProvider>
		);
	});
	it("should render the component", () => {
		expect(wrapper.containsMatchingElement(<GameControls />)).toEqual(true);
	});
	it("should match the snapshot", () => {
		expect(wrapper).toMatchSnapshot();
	});
});
