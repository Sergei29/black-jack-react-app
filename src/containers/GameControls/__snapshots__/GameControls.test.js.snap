// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameControls should match the snapshot 1`] = `ReactWrapper {}`;
